### ISSUE TITLE
viewer dependencies (freetype-py & pyglfw) do not include arm binaries for m1 macs

### ISSUE THREAD

**Issue by Mark2Mark** [coldtype/coldtype#70]
Thanks so much for looking into #27 @stenson! I really appreciate all your help.

On a follow up on that:
I could now get a few steps further, however I had to upgrade pip after it threw a ton of new errors at me. These are for later reference, perhaps this could be added to the install instructions.

either this `pip install --upgrade pip setuptools wheel`
 or into the venv: `{...}/venv/bin/python3.9 -m pip install --upgrade pip` not sure, I used my absolute path with was proposed by the terminal error. It seems that the PIP inside of the venv always needs to be updated. I tried again after I updated my local pip, and it threw the error about an outdated pip again.
 
 then I do the `pip install "coldtype[viewer]"`  and can’t see any errors in the log.
 
 However, if trying to continue with `coldtype demo`
 
 I get a bunch of errors again
 
 ```sh
Traceback (most recent call last):
  File "/{MY_PATH}/Coldtype/venv/lib/python3.9/site-packages/freetype/raw.py", line 49, in <module>
    _lib = ctypes.CDLL(filename)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/ctypes/__init__.py", line 374, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: dlopen(libfreetype.dylib, 6): image not found

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/{MY_PATH}/Coldtype/venv/bin/coldtype", line 5, in <module>
    from coldtype.renderer import main
  File "/{MY_PATH}/Coldtype/venv/lib/python3.9/site-packages/coldtype/__init__.py", line 4, in <module>
    from coldtype.text import *
  File "/{MY_PATH}/Coldtype/venv/lib/python3.9/site-packages/coldtype/text/__init__.py", line 3, in <module>
    import coldtype.text.composer
  File "/{MY_PATH}/Coldtype/venv/lib/python3.9/site-packages/coldtype/text/composer.py", line 7, in <module>
    from coldtype.text.reader import Style, StyledString, FittableMixin, Font, normalize_font_path, SegmentedString,  _PenClass, _PensClass
  File "/{MY_PATH}/Coldtype/venv/lib/python3.9/site-packages/coldtype/text/reader.py", line 42, in <module>
    from coldtype.fontgoggles.font.otfFont import OTFFont
  File "/{MY_PATH}/Coldtype/venv/lib/python3.9/site-packages/coldtype/fontgoggles/font/otfFont.py", line 7, in <module>
    from ..misc.ftFont import FTFont
  File "/{MY_PATH}/Coldtype/venv/lib/python3.9/site-packages/coldtype/fontgoggles/misc/ftFont.py", line 5, in <module>
    import freetype
  File "/{MY_PATH}/Coldtype/venv/lib/python3.9/site-packages/freetype/__init__.py", line 24, in <module>
    from freetype.raw import *
  File "/{MY_PATH}/Coldtype/venv/lib/python3.9/site-packages/freetype/raw.py", line 52, in <module>
    raise RuntimeError('Freetype library not found')
RuntimeError: Freetype library not found
 ```

freetype is installed, running `pip install freetype-py` claims `Requirement already satisfied: freetype-py in ./venv/lib/python3.9/site-packages (2.2.0)`. It also shows in the list `pip list -l` as freetype-py 2.2.0

_Originally posted by @Mark2Mark in https://github.com/goodhertz/coldtype/issues/27#issuecomment-945078202_

**Comment by Mark2Mark** [comment]
Here’s how it’s going.

I reinstalled freetype with brew, and located it in
`/opt/homebrew/Cellar/freetype/2.11.0/lib/libfreetype.6.dylib`

Then I copied that file to the venv `.../venv/lib/python3.9/site-packages/freetype/libfreetype.dylib` and renamed it to `libfreetype.dylib`

Now I don’t get any errors anymore on `coldtype demo`

But now it is stuck on `... watching venv/lib/python3.9/site-packages/coldtype/demo/demo.py for changes ...` while no window appears as expected. No python app in the Dock.

**Comment by stenson** [comment]
Hi Mark — sorry for the all the trouble here! Unfortunately I haven't seen any of these errors before and I don't have access to an M1 computer, although I know of at least one M1 computer that has successfully run an older version of coldtype, so I'm not sure what the issue there would be. Could definitely be M1, though, since the prebuilt binaries that some python libraries come with might not be built properly for M1, which is what the freetype-py issue sounds like (``libfreetype.dylib: mach-o, but wrong architecture``) except that I can't find anything online about people freetype-py doesn't work on M1 which seems a little odd.

What OS version are you running?

And if you are comfortable modifying the coldtype source, if you locate the `coldtype/renderer/__init__.py` file in your venv, could you try modifying the `initialize_gui_and_server` function to have an additional print at the beginning, so it'd look something like this:

```python
    def initialize_gui_and_server(self):
        self.winmans.add_viewers()
        
        print("GLSK >>>", self.winmans.glsk)
```

The `glsk` there is the viewer (a GLFW window), so if that's `None` then it's not being created, which could happen if the `skia-python` or `glfw` libraries aren't being installed properly.

**Comment by Mark2Mark** [comment]
Hi @stenson Thanks so much for your quick reply.

I think we’re pretty close, as the M1 issue was apparently semi-solved by me dumping the latest dylib in there myself. (I think you could bundle the newest freetype as it ships [from here](https://formulae.brew.sh/formula/freetype) (that’s the one I uploaded as a zip in my comment above). There it’s mentioned to be compatible with Silicon and Big Sur.

I was happy to add your debug print, however it prints `GLSK >>> None`

Can I somehow install skia-python and or glfw in a way that coldtype can use it?

Thanks so so much for all your help, I am still so excited to get it running. Never saw it on my screen yet :D

**Comment by Mark2Mark** [comment]
Thanks so much.

I couldn’t find the version, but I ran this pip insall and it seems it was already that version:
`Requirement already satisfied: glfw==2.3.0 in ./venv/lib/python3.9/site-packages (2.3.0)`

**Comment by stenson** [comment]
Hmm, yeah I guess that should have the fix then. Do you know where glfw is installed to in your homebrew?

**Comment by Mark2Mark** [comment]
I don’t have it installed in homebrew (generally I try to not use homebrew anyway, just got it for a necessary web dev project).

But I did that build as mentioned here
> Now I was building a GLFW with the [commands as state here](https://gist.github.com/vaiorabbit/6204931) and got a folder with a build that might work (I can run the example files in `{...}/glfw-3.3.2/build/examples`).
which just sits in a folder. If I knew if I could pick a few files out of that and dump it into the venv …

Ah, so sory for that …

**Comment by Mark2Mark** [comment]
I did it now with the Homebrew one and got a tiny bit futher with a next error now.
Then I found the same-named libglfw in the arm build (again see see my quoted part above) and it shows the same error.

But I got the feeling we’re closer

```
❯ coldtype demo
... watching venv/lib/python3.9/site-packages/coldtype/demo/demo.py for changes ...
GLSK >>> <coldtype.renderer.winman.glfwskia.WinmanGLFWSkia object at 0x110427850>
[1]    62709 segmentation fault  coldtype demo
```

After the error, an NSAlert pops up, telling me that "Python quit unexpectedly."

Thank *you* so much for sticking with me, I know how annoying it is to debug such things, especially remotely. I really appreciate all your help!

**Comment by Mark2Mark** [comment]
Thank you all for your help. Unfortunately now I don’t know if it was just this one little line as mentioned in my last answer. I will try to start from scratch tomorrow and see if that already solves it. Otherwise I’ll go through the steps and make a list here of M1 requirements. It could be that the homebrew and/or self-built dylib was also necesary.

**Comment by stenson** [comment]
@Mark2Mark great to hear it's working! My guess is that `pin and False` might be a big part of it, although wouldn't solve the needing-to-homebrew-install for freetype & glfw

@jo-lang thanks for chiming in! very helpful to get another m1 perspective in here — did you have rosetta enabled when you built the virtualenv?

**Comment by Mark2Mark** [comment]
Before turning the gathered info about dependencies into a proper list, i just quickly dump more things here:
- FFmpeg `brew install ffmpeg`
- ...

**Comment by FlorianRhiem** [comment]
I've just released version 2.4.0 of the package `glfw`, along with wheels for macOS with arm64, so the pyglfw half of this issue should be fixed.

I'm using the binaries provided by GLFW3 and do not have an M1 macbook myself, so if there are any issues with the wheels, please let me know!

**Comment by stenson** [comment]
@FlorianRhiem thanks for letting us know! working great on my m1 laptop with an arm python3.9
